File: examples/go/ttl-caching/collection_management.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"

	"github.com/joho/godotenv"
)

func main() {
	// Load environment variables from the .env file
	err := godotenv.Load()
	if err != nil {
		log.Fatal(fmt.Sprintf("✗ Failed to load environment variables: %v", err))
		os.Exit(1)
	}

	var baseURL = os.Getenv("API_BASE_URL")
	var apiKey = os.Getenv("API_BASE_KEY")

	// Get authentication token
	token := os.Getenv("AUTH_TOKEN")
	if token == "" {
		tokenResp, err := http.Post(baseURL+"/api/auth/token", "application/json",
			bytes.NewBuffer([]byte(fmt.Sprintf(`{"api_key":"%s"}`, apiKey))))
		if err != nil {
			log.Fatal(err)
		}
		defer tokenResp.Body.Close()

		var tokenData map[string]interface{}
		json.NewDecoder(tokenResp.Body).Decode(&tokenData)
		token = tokenData["token"].(string)
	}

	fmt.Println("✓ Authentication successful")

	client := &http.Client{}

	// Example 1: Create a collection (via insert)
	fmt.Println("\n=== Create Collection (via insert) ===")
	insertData := map[string]interface{}{
		"name":        "Demo Collection",
		"description": "Created for testing",
		"active":      true,
	}
	insertJSON, _ := json.Marshal(insertData)

	req, _ := http.NewRequest("POST", baseURL+"/api/insert/demo_collection", bytes.NewBuffer(insertJSON))
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	resp, _ := client.Do(req)
	var insertResult map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&insertResult)
	resp.Body.Close()
	fmt.Printf("Collection created with first record: %s\n", insertResult["id"])

	// Example 2: List all collections
	fmt.Println("\n=== List Collections ===")
	req, _ = http.NewRequest("GET", baseURL+"/api/collections", nil)
	req.Header.Set("Authorization", "Bearer "+token)

	resp, _ = client.Do(req)
	var collectionsData map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&collectionsData)
	resp.Body.Close()

	collections := collectionsData["collections"].([]interface{})
	fmt.Printf("Total collections: %d\n", len(collections))
	fmt.Printf("Sample collections: %v\n", collections)

	// Example 3: Count documents in collection
	fmt.Println("\n=== Count Documents ===")
	countData := map[string]interface{}{
		"limit": 1000,
	}
	countJSON, _ := json.Marshal(countData)

	req, _ = http.NewRequest("POST", baseURL+"/api/find/demo_collection", bytes.NewBuffer(countJSON))
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	resp, _ = client.Do(req)
	var docs []interface{}
	json.NewDecoder(resp.Body).Decode(&docs)
	resp.Body.Close()
	fmt.Printf("Document count: %d\n", len(docs))

	// Example 4: Delete collection
	fmt.Println("\n=== Delete Collection ===")
	req, _ = http.NewRequest("DELETE", baseURL+"/api/collections/demo_collection", nil)
	req.Header.Set("Authorization", "Bearer "+token)

	resp, _ = client.Do(req)
	resp.Body.Close()
	fmt.Println("Collection deleted successfully")

	// Example 5: Verify deletion
	fmt.Println("\n=== Verify Deletion ===")
	req, _ = http.NewRequest("GET", baseURL+"/api/collections", nil)
	req.Header.Set("Authorization", "Bearer "+token)

	resp, _ = client.Do(req)
	var updatedCollectionsData map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&updatedCollectionsData)
	resp.Body.Close()

	updatedCollections := updatedCollectionsData["collections"].([]interface{})
	stillExists := false
	for _, col := range updatedCollections {
		if col.(string) == "demo_collection" {
			stillExists = true
			break
		}
	}
	fmt.Printf("Collection still exists: %v\n", stillExists)

	fmt.Println("\n✓ All collection management operations completed successfully")
}

File: examples/rust/examples/collection_management.rs
use serde_json::{json, Value};
use std::env;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    // Step 1: Get authentication token
    let client = reqwest::Client::new();
    let token_response = client
        .post(&format!("{}/api/auth/token", base_url))
        .json(&json!({ "api_key": api_key }))
        .send()
        .await?;

    let token_data: Value = token_response.json().await?;
    let token = token_data["token"].as_str().unwrap();
    println!("✓ Authentication successful");

    // Example 1: Create a collection (via insert)
    println!("\n=== Create Collection (via insert) ===");
    let insert_data = json!({
        "name": "Demo Collection",
        "description": "Created for testing",
        "active": true
    });

    let insert_response = client
        .post(&format!("{}/api/insert/demo_collection", base_url))
        .header("Authorization", format!("Bearer {}", token))
        .json(&insert_data)
        .send()
        .await?;

    let insert_result: Value = insert_response.json().await?;
    println!(
        "Collection created with first record: {}",
        insert_result["id"]
    );

    // Example 2: List all collections
    println!("\n=== List Collections ===");
    let collections_response = client
        .get(&format!("{}/api/collections", base_url))
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await?;

    let collections_data: Value = collections_response.json().await?;
    let collections = collections_data["collections"].as_array().unwrap();
    println!("Total collections: {}", collections.len());

    let collection_names: Vec<String> = collections
        .iter()
        .filter_map(|v| v.as_str().map(|s| s.to_string()))
        .collect();
    println!("Sample collections: {:?}", collection_names);

    // Example 3: Count documents in collection
    println!("\n=== Count Documents ===");
    let count_response = client
        .post(&format!("{}/api/find/demo_collection", base_url))
        .header("Authorization", format!("Bearer {}", token))
        .json(&json!({"limit": 1000}))
        .send()
        .await?;

    let docs: Vec<Value> = count_response.json().await?;
    println!("Document count: {}", docs.len());

    // Example 4: Delete collection
    println!("\n=== Delete Collection ===");
    client
        .delete(&format!("{}/api/collections/demo_collection", base_url))
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await?;

    println!("Collection deleted successfully");

    // Example 5: Verify deletion
    println!("\n=== Verify Deletion ===");
    let verify_response = client
        .get(&format!("{}/api/collections", base_url))
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await?;

    let updated_collections_data: Value = verify_response.json().await?;
    let updated_collections = updated_collections_data["collections"].as_array().unwrap();
    let updated_collection_names: Vec<String> = updated_collections
        .iter()
        .filter_map(|v| v.as_str().map(|s| s.to_string()))
        .collect();
    let still_exists = updated_collection_names.contains(&"demo_collection".to_string());
    println!("Collection still exists: {}", still_exists);

    println!("\n✓ All collection management operations completed successfully");

    Ok(())
}
